### PR TITLE
Fix writing and reading extension type

### LINKED ISSUE #170: Segfault reading converted coco dataset
Repro:

Branch: changhiskhan/coco

Note: the raw data uri needs to be updated for local system:
```python
from parse_coco import CocoConverter
c = CocoConverter('/home/chang/Downloads/datasets/coco')
df = c.read_metadata()
table = c.save_df(df, output_path='/tmp/coco_test.lance')

import lance
lance.dataset('/tmp/coco_test.lance').head(1)
```

Segfault:

```
./cpp/src/arrow/array/array_nested.cc:191:  Check failed: (self->list_type_->value_type()->id()) == (data->child_data[0]->type->id()) 
/lib/x86_64-linux-gnu/libarrow.so.900(+0x14c16ef)[0x7f36220cd6ef]
/lib/x86_64-linux-gnu/libarrow.so.900(_ZN5arrow4util8ArrowLogD1Ev+0xe1)[0x7f36219b81b1]
/lib/x86_64-linux-gnu/libarrow.so.900(+0xf6969b)[0x7f3621b7569b]
/lib/x86_64-linux-gnu/libarrow.so.900(_ZN5arrow9ListArrayC1ESt10shared_ptrINS_8DataTypeEElS1_INS_6BufferEES1_INS_5ArrayEES5_ll+0x183)[0x7f3621b760d3]
/home/chang/code/eto/lance/cpp/build/liblance.so(_ZSt12construct_atIN5arrow9ListArrayEJSt10shared_ptrINS0_8DataTypeEElRS2_INS0_6BufferEERS2_INS0_5ArrayEES7_EEDTgsnwcvPvLi0E_T_pispcl7declvalIT0_EEEEPSC_DpOSD_+0x13f)[0x7f35f60247db]
/home/chang/code/eto/lance/cpp/build/liblance.so(_ZNSt16allocator_traitsISaIN5arrow9ListArrayEEE9constructIS1_JSt10shared_ptrINS0_8DataTypeEElRS5_INS0_6BufferEERS5_INS0_5ArrayEESA_EEEvRS2_PT_DpOT0_+0x8b)[0x7f35f6024912]
/home/chang/code/eto/lance/cpp/build/liblance.so(_ZNSt23_Sp_counted_ptr_inplaceIN5arrow9ListArrayESaIS1_ELN9__gnu_cxx12_Lock_policyE2EEC1IJSt10shared_ptrINS0_8DataTypeEElRS7_INS0_6BufferEERS7_INS0_5ArrayEESC_EEES2_DpOT_+0x10f)[0x7f35f6022cad]
/home/chang/code/eto/lance/cpp/build/liblance.so(_ZNSt14__shared_countILN9__gnu_cxx12_Lock_policyE2EEC2IN5arrow9ListArrayESaIS5_EJSt10shared_ptrINS4_8DataTypeEElRS7_INS4_6BufferEERS7_INS4_5ArrayEESC_EEERPT_St20_Sp_alloc_shared_tagIT0_EDpOT1_+0x157)[0x7f35f602124f]
/home/chang/code/eto/lance/cpp/build/liblance.so(_ZNSt12__shared_ptrIN5arrow9ListArrayELN9__gnu_cxx12_Lock_policyE2EEC2ISaIS1_EJSt10shared_ptrINS0_8DataTypeEElRS7_INS0_6BufferEERS7_INS0_5ArrayEESC_EEESt20_Sp_alloc_shared_tagIT_EDpOT0_+0xa4)[0x7f35f601f7f0]
/home/chang/code/eto/lance/cpp/build/liblance.so(_ZNSt10shared_ptrIN5arrow9ListArrayEEC2ISaIS1_EJS_INS0_8DataTypeEElRS_INS0_6BufferEERS_INS0_5ArrayEES9_EEESt20_Sp_alloc_shared_tagIT_EDpOT0_+0x97)[0x7f35f601e0db]
/home/chang/code/eto/lance/cpp/build/liblance.so(_ZSt15allocate_sharedIN5arrow9ListArrayESaIS1_EJSt10shared_ptrINS0_8DataTypeEElRS3_INS0_6BufferEERS3_INS0_5ArrayEES8_EES3_IT_ERKT0_DpOT1_+0x91)[0x7f35f601b9f6]
/home/chang/code/eto/lance/cpp/build/liblance.so(_ZSt11make_sharedIN5arrow9ListArrayEJSt10shared_ptrINS0_8DataTypeEElRS2_INS0_6BufferEERS2_INS0_5ArrayEES7_EES2_IT_EDpOT0_+0xb4)[0x7f35f60184c8]
/home/chang/code/eto/lance/cpp/build/liblance.so(_ZNK5lance2io10FileReader12GetListArrayERKSt10shared_ptrINS_6format5FieldEEiRKNS1_15ArrayReadParamsE+0xde6)[0x7f35f6012e9e]
/home/chang/code/eto/lance/cpp/build/liblance.so(_ZNK5lance2io10FileReader8GetArrayERKSt10shared_ptrINS_6format5FieldEEiRKNS1_15ArrayReadParamsE+0x11e)[0x7f35f6011922]
/home/chang/code/eto/lance/cpp/build/liblance.so(_ZNK5lance2io10FileReader14GetStructArrayERKSt10shared_ptrINS_6format5FieldEEiRKNS1_15ArrayReadParamsE+0xf9)[0x7f35f6011e1b]
/home/chang/code/eto/lance/cpp/build/liblance.so(_ZNK5lance2io10FileReader8GetArrayERKSt10shared_ptrINS_6format5FieldEEiRKNS1_15ArrayReadParamsE+0xb4)[0x7f35f60118b8]
/home/chang/code/eto/lance/cpp/build/liblance.so(_ZNK5lance2io10FileReader9ReadBatchERKNS_6format6SchemaEiRKNS1_15ArrayReadParamsE+0xdf)[0x7f35f60113d7]
/home/chang/code/eto/lance/cpp/build/liblance.so(_ZNK5lance2io10FileReader9ReadBatchERKNS_6format6SchemaEiiSt8optionalIiE+0x66)[0x7f35f60111cc]
/home/chang/code/eto/lance/cpp/build/liblance.so(_ZN5lance2io4exec4Scan4NextEv+0x21b)[0x7f35f603eb9b]
/home/chang/code/eto/lance/cpp/build/liblance.so(_ZN5lance2io4exec5Limit4NextEv+0xee)[0x7f35f603aefc]
/home/chang/code/eto/lance/cpp/build/liblance.so(_ZN5lance2io4exec7Project4NextEv+0x85)[0x7f35f603ce5f]
/home/chang/code/eto/lance/cpp/build/liblance.so(_ZNK5lance2io17RecordBatchReader9ReadBatchEv+0x56)[0x7f35f602d2f2]
/home/chang/code/eto/lance/cpp/build/liblance.so(+0x366408)[0x7f35f602d408]
/home/chang/code/eto/lance/cpp/build/liblance.so(+0x36745c)[0x7f35f602e45c]
/home/chang/code/eto/lance/cpp/build/liblance.so(+0x367393)[0x7f35f602e393]
/home/chang/code/eto/lance/cpp/build/liblance.so(+0x3672d3)[0x7f35f602e2d3]
/home/chang/code/eto/lance/cpp/build/liblance.so(+0x367212)[0x7f35f602e212]
/home/chang/code/eto/lance/cpp/build/liblance.so(+0x367177)[0x7f35f602e177]
/home/chang/code/eto/lance/cpp/build/liblance.so(+0x367140)[0x7f35f602e140]
/lib/x86_64-linux-gnu/libarrow.so.900(+0xda9c90)[0x7f36219b5c90]
/lib/x86_64-linux-gnu/libstdc++.so.6(+0xdc2b3)[0x7f36209732b3]
/lib/x86_64-linux-gnu/libc.so.6(+0x94b43)[0x7f362681cb43]
/lib/x86_64-linux-gnu/libc.so.6(+0x126a00)[0x7f36268aea00]
[1]    686511 IOT instruction (core dumped)  ipython
```

The problem seems to be that Lance is not writing the nested extension type properly:

```python
from parse_coco import CocoConverter
c = CocoConverter('/home/chang/Downloads/datasets/coco')
df = c.read_metadata()
table = c._convert_metadata_df(df)
table.schema
```

The above shows that `annotations.bbox` has type `list<box2d>` which is as expected.

But once it's written, only the storage type is read:

```python
lance.dataset('/tmp/coco_test.lance').schema
``` 

The above shows that `annotations.bbox` is `list<fixed_size_list<double>>`

This is consistent with the segfault stack trace above showing:

```
./cpp/src/arrow/array/array_nested.cc:191:  Check failed: (self->list_type_->value_type()->id()) == (data->child_data[0]->type->id()) 
```

### PR DESCRIPTION
**Pull request by eddyxu**
Closes #170

### CODE REVIEW

File: cpp/src/lance/format/schema.cc [hunk 1]
```diff
@@ -46,14 +46,15 @@ Field::Field(const std::shared_ptr<::arrow::Field>& field)
       extension_name_(arrow::GetExtensionName(field->type()).value_or("")),
       encoding_(pb::NONE) {
   if (is_extension_type()) {
-    auto ext_type = std::static_pointer_cast<::arrow::ExtensionType>(field->type());
+    auto ext_type = std::dynamic_pointer_cast<::arrow::ExtensionType>(field->type());
```
Comment (changhiskhan) [inline cpp/src/lance/format/schema.cc:49]: the dynamic cast is for subclasses of ExtensionType right?

File: cpp/src/lance/format/schema.cc [hunk 2]
```diff
@@ -185,7 +183,8 @@ ::arrow::Status Field::set_dictionary(std::shared_ptr<::arrow::Array> dict_arr)
 }
 
 ::arrow::Status Field::LoadDictionary(std::shared_ptr<::arrow::io::RandomAccessFile> infile) {
-  assert(::arrow::is_dictionary(type()->id()));
+  auto data_type = storage_type();
```
Comment (changhiskhan) [inline cpp/src/lance/format/schema.cc:186]: where is data_type used?

File: cpp/src/lance/format/schema.cc [hunk 3]
```diff
@@ -224,7 +223,7 @@ std::shared_ptr<lance::encodings::Encoder> Field::GetEncoder(
 ::arrow::Result<std::shared_ptr<lance::encodings::Decoder>> Field::GetDecoder(
     std::shared_ptr<::arrow::io::RandomAccessFile> infile) {
   std::shared_ptr<lance::encodings::Decoder> decoder;
-  auto data_type = type();
+  auto data_type = storage_type();
```
Comment (changhiskhan) [inline cpp/src/lance/format/schema.cc:226]: not sure where this should go in code comments / docs but let's formalize the abstraction that the encoder/decoder only deals with storage types when reading/writing data.

File: cpp/src/lance/format/schema.cc [hunk 4]
```diff
@@ -327,6 +323,14 @@ std::shared_ptr<::arrow::DataType> Field::type() const {
   }
 }
 
+std::shared_ptr<::arrow::DataType> Field::storage_type() const {
```
Comment (changhiskhan) [inline cpp/src/lance/format/schema.cc:326]: how come in `type()` we expand nested types but not in `storage_type()` ?

Comment (changhiskhan) [reply cpp/src/lance/format/schema.cc:326]: e.g., if you have a nested type with extension types in subfields, `storage_type()` won't actually convert the subfield extension types right?

File: cpp/src/lance/io/writer.h [hunk 5]
```diff
@@ -53,8 +53,12 @@ class FileWriter final : public ::arrow::dataset::FileWriter {
 
   ::arrow::Status WriteArray(const std::shared_ptr<format::Field>& field,
                              const std::shared_ptr<::arrow::Array>& arr);
-  ::arrow::Status WritePrimitiveArray(const std::shared_ptr<format::Field>& field,
-                                      const std::shared_ptr<::arrow::Array>& arr);
+
+  /// Write fixed size array, including primitive arrays, fixed size binary array,
+  /// and fixed size list array.
+  ::arrow::Status WriteFixedSizedArray(const std::shared_ptr<format::Field>& field,
```
Comment (changhiskhan) [inline cpp/src/lance/io/writer.h:59]: it's a little confusing that this name makes you think it's dealing with fixed sized list / fixed sized binary only. If it's not too much a hassle, can we consider renaming this to something more generic like "WriteFixedStrideArray" or something like that?

File: cpp/src/lance/format/schema.cc (hunk 1 above)
Comment (eddyxu) [reply cpp/src/lance/format/schema.cc:49]: `field->type()` returns `std::shared_ptr<DataType>`. `ExtensionType` is a subclass of DataType, so use dynamic cast here?

File: cpp/src/lance/io/writer.h (hunk 5 above)
Comment (eddyxu) [reply cpp/src/lance/io/writer.h:59]: Sure , can do

File: cpp/src/lance/format/schema.cc (hunk 4 above)
Comment (eddyxu) [reply cpp/src/lance/format/schema.cc:326]: Good point. will fix it.

File: cpp/src/lance/io/writer.h (hunk 5 above)
Comment (changhiskhan) [reply cpp/src/lance/io/writer.h:59]: yeah that sounds fine

File: cpp/src/lance/format/schema.cc (hunk 2 above)
Comment (eddyxu) [reply cpp/src/lance/format/schema.cc:186]: Missed the usage. Fixed.

File: cpp/src/lance/format/schema.cc (hunk 3 above)
Comment (eddyxu) [reply cpp/src/lance/format/schema.cc:226]: Added a comment to `Decoder` base class to mention that it does not accept ExtensionType

**Review by changhiskhan — APPROVED** [review]

**Pull request merged**